### PR TITLE
Restructure the model

### PR DESCRIPTION
**Pull request by ftsamis**
**Note: Closed in favour of #632, in order to change the target branch**

The purpose of this PR is to rewrite the `Radial1DModel` class to make it more dynamic and add a proper `__init__` method to it, which accepts all the required arguments for the model. Also, attributes and methods that do not belong to the model should be removed from it with this PR.

The main idea is to have some base attributes to the model, and then provide dynamic arguments (i.e. python properties) which are (re)calculated every time they are accessed from the base attributes. This way, if a base attribute is changed, everything that depends on it will reflect the changes.

Additionally, `__init__` should now take the actual base attributes that model needs instead of a config object, while we will still have a `from_config` classmethod which will create a model from a config object.

### CODE REVIEW

File: tardis/model/base.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class Radial1DModel(object):
+    def __init__(self, plasma, atom_data, structure):
+        pass
+
+    @classmethod
+    def from_config(self, config):
```
Comment (wkerzendorf) [inline tardis/model/base.py]: @ftsamis velocities, densities, v_inner, t_inner, abundances

Comment (ftsamis) [inline tardis/model/base.py]: Does that include plasma, atom_data, structure?

Comment (ftsamis) [inline tardis/model/base.py]: @wkerzendorf What do you think about: `__init__(self, plasma, atom_data, velocities, densities, structure, abundances)`

`t_inner` could be read from `plasma.t_inner`,
`v_inner` and `v_outer` could be read from `structure` which could be a new object.

Does `velocities` have anything to do with `v_inner` or `v_outer`?

Comment (yeganer) [inline tardis/model/base.py]: `velocities == np.hstack([v_inner, v_outer[:-1])`
It's something like that. So `velocities` has the same information as `v_inner` and `v_outer` combined.

File: tardis/model/base.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class Radial1DModel(object):
+    def __init__(self, plasma, atom_data, structure):
```
Comment (wkerzendorf) [inline tardis/model/base.py]: okay the init is going to have `velocity`, `density`, `abundance`, `t_rad`, `w`(optional defaults to None), `v_boundary_inner`, `v_boundary_outer`.

File: tardis/model/base.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+class Radial1DModel(object):
+    def __init__(self, velocity, density, abundance, t_rad,
```
Comment (wkerzendorf) [inline tardis/model/base.py]: `time_explosion`

Comment (wkerzendorf) [inline tardis/model/base.py]: `time_density`

Comment (wkerzendorf) [inline tardis/model/base.py]: `t_radiative`, `dilution_factor=None`

File: tardis/model/base.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+                 v_boundary_inner, v_boundary_outer, w=None):
+        self.velocity = velocity
+        self.density = density
+        self.abundance = abundance
+        self.t_rad = t_rad
+        self.v_boundary_inner = v_boundary_inner
+        self.v_boundary_outer = v_boundary_outer
+        self.w = w
```
Comment (wkerzendorf) [inline tardis/model/base.py]:
```
@property
def w(self):
    return self.dilution_factor
```

File: tardis/model/base.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class Radial1DModel(object):
+    def __init__(self, velocity, density, abundance, t_rad,
+                 v_boundary_inner, v_boundary_outer, w=None):
+        self.velocity = velocity
+        self.density = density
+        self.abundance = abundance
+        self.t_rad = t_rad
```
Comment (wkerzendorf) [inline tardis/model/base.py]:
```
@property
def t_rad(self):
    return self.t_radiative
```

Comment (wkerzendorf) [inline tardis/model/base.py]: descriptors.

File: tardis/model/base.py (hunk 3 above)
Comment (wkerzendorf) [inline tardis/model/base.py]: `density_0`

File: tardis/model/base.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+class Radial1DModel(object):
```
Comment (wkerzendorf) [inline tardis/model/base.py]: add a new density object.

```
class HomologousDensity(object):
    def __init__(self, density_0, time_0):
        self.density_0 = density_0
        self.time_0 = time_0
        self.time
     @property
     def density(self):
          return (t/t_0)**-3 * self.density_0
```

File: tardis/model/base.py (hunk 2 above)
Comment (ftsamis) [inline tardis/model/base.py]: Done

File: tardis/model/base.py (hunk 3 above)
Comment (ftsamis) [inline tardis/model/base.py]: Done

File: tardis/model/base.py (hunk 4 above)
Comment (ftsamis) [inline tardis/model/base.py]: Done

File: tardis/model/base.py (hunk 5 above)
Comment (ftsamis) [inline tardis/model/base.py]: Done: getter/setter alias t_rad for t_radiative

File: tardis/model/base.py (hunk 6 above)
Comment (ftsamis) [inline tardis/model/base.py]: In progress. Subclassing `Quantity` is possible, but not simple.

Comment (yeganer) [inline tardis/model/base.py]: @wkerzendorf What's the plan for `config.structure.volumes`? Where should that live? To me it feels right to put it into the model, too.

Comment (ftsamis) [inline tardis/model/base.py]: Regarding the density, apparently descriptors are not the right option here, a descriptor must be a class attribute, and not an instance attribute.

So it would look something like this:
```python
class Model(object):
    density = HomologousDensity(density_0=xyz, time_0=xyz)
    def __init__(self, ...):
        ....
```

which is not what we want, since we need `density_0` and `time_0` to be different for each model instance. If I put the density assignment line inside the init method, the descriptor will not work.

File: tardis/model/base.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
+import numpy as np
+
+
+class Density(object):
```
Comment (wkerzendorf) [inline tardis/model/base.py]: `HomologousDensity`

File: tardis/model/base.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    def __init__(self, density_0, time_0):
+        self.density_0 = density_0
+        self.time_0 = time_0
+
+    def after_time(self, time_explosion):
+        return self.density_0 * (time_explosion / self.time_0) ** -3
+
+
+class Radial1DModel(object):
+    def __init__(self, velocity, density, abundance, t_inner, t_radiative,
```
Comment (wkerzendorf) [inline tardis/model/base.py]: `homologous_density`

**Comment by ftsamis** [comment]
I have no idea where to start about this but let's see...

I've now added a `model/density.py` module holding one class for each type of density mode we currently support. The classes there are actually special descriptors, designed to be used with [a hack](http://blog.brianbeck.com/post/74086029/instance-descriptors) which makes descriptors work in instances (instead of needing to be class attributes).

Most of the densities (except uniform) depend on the model's velocity (`v_middle`). By using descriptors we can have access to the model velocity and calculate the density on the fly, which means that the model can change entirely, and the density will "automatically adapt" to reflect the new changes (it will just be recalculated each time accessed).

The same thing happens with `time_explosion`. `model.density` will always have the scaled to the correct time density.

Of course this needs a serious amount of reviewing and testing.

Note 1: I have not yet removed the code that this PR replaces in `model/legacy` and `config_reader`, I will do that soon.

Note 2: I haven't added full support for `v_boundaries`, yet, but I have most of the prerequisites there such as `v_boundary_inner_index` and `v_boundary_outer_index`, which I would appreciate it if you reviewed them too.

TL;DR version:

```python
from astropy.units import Quantity

import tardis
from tardis.model.base import Radial1DModel as NewModel # Import the brand-new model class
from tardis.io.config_reader import Configuration

# tardis_example.yml uses branch85_w7 density
config = Configuration.from_yaml('tardis_example.yml')
mdl = NewModel.from_config(config)
print "Initial state"
print mdl.density
# Compare the density as calculated by the descriptors to the old reference, before we start messing with it
print mdl.density == config.structure.mean_densities

# Change the time_explosion and see how density adapts
time_orig = mdl.time_explosion
mdl.time_explosion = mdl.time_explosion * 1.5
print "time_explosion increased by 50%"
print mdl.density

# Reset the time_explosion
print "time_explosion reset"
mdl.time_explosion = time_orig
print mdl.density

# Mess with the velocity array and see how density adapts
print "velocity decreased by 20%"
velocity_orig = mdl.velocity
mdl.velocity = mdl.velocity * 0.8
print mdl.density

# Reset the time_explosion
print "velocity reset"
mdl.velocity = velocity_orig
print mdl.density
```

```
    tardis.io.config_reader - INFO - Reading Atomic Data from kurucz_cd23_chianti_H_He.h5
    tardis.atomic - INFO - Read Atom Data with UUID=5ca3035ca8b311e3bb684437e69d75d7 and MD5=21095dd25faa1683f4c90c911a00c3f8
    tardis.io.config_reader - INFO - "initial_t_inner" is not specified in the plasma section - initializing to 9933.95199592 K with given luminosity
    tardis.io.config_reader - WARNING - No "species" given - ignoring other NLTE options given:
    OrderedDict([('coronal_approximation', False), ('classical_nebular', False)])


    Initial state
    [  7.54280360e-14   5.72847501e-14   4.39607422e-14   3.40628743e-14
       2.66313464e-14   2.09959645e-14   1.66828721e-14   1.33531053e-14
       1.07615377e-14   8.72908485e-15   7.12365156e-15   5.84692095e-15
       4.82509279e-15   4.00232416e-15   3.33603855e-15   2.79354038e-15
       2.34955043e-15   1.98439683e-15   1.68267689e-15   1.43225980e-15] g / cm3

    [ True  True  True  True  True  True  True  True  True  True  True  True
      True  True  True  True  True  True  True  True]

    time_explosion increased by 50%
    [  2.23490477e-14   1.69732593e-14   1.30254051e-14   1.00927035e-14
       7.89076930e-15   6.22102653e-15   4.94307322e-15   3.95647564e-15
       3.18860378e-15   2.58639551e-15   2.11071157e-15   1.73242102e-15
       1.42965712e-15   1.18587383e-15   9.88455867e-16   8.27715669e-16
       6.96163091e-16   5.87969432e-16   4.98570929e-16   4.24373274e-16] g / cm3

    time_explosion reset
    [  7.54280360e-14   5.72847501e-14   4.39607422e-14   3.40628743e-14
       2.66313464e-14   2.09959645e-14   1.66828721e-14   1.33531053e-14
       1.07615377e-14   8.72908485e-15   7.12365156e-15   5.84692095e-15
       4.82509279e-15   4.00232416e-15   3.33603855e-15   2.79354038e-15
       2.34955043e-15   1.98439683e-15   1.68267689e-15   1.43225980e-15] g / cm3

    velocity decreased by 20%
    [  3.59668903e-13   2.73154974e-13   2.09621154e-13   1.62424442e-13
       1.26988155e-13   1.00116561e-13   7.95501333e-14   6.36725679e-14
       5.13150107e-14   4.16235201e-14   3.39682177e-14   2.78802917e-14
       2.30078353e-14   1.90845688e-14   1.59074714e-14   1.33206386e-14
       1.12035295e-14   9.46234147e-15   8.02362865e-15   6.82954692e-15] g / cm3

    velocity reset
    [  7.54280360e-14   5.72847501e-14   4.39607422e-14   3.40628743e-14
       2.66313464e-14   2.09959645e-14   1.66828721e-14   1.33531053e-14
       1.07615377e-14   8.72908485e-15   7.12365156e-15   5.84692095e-15
       4.82509279e-15   4.00232416e-15   3.33603855e-15   2.79354038e-15
       2.34955043e-15   1.98439683e-15   1.68267689e-15   1.43225980e-15] g / cm3
```

File: tardis/model/base.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        if t_radiative is None:
+            lambda_wien_inner = constants.b_wien / self.t_inner
+            self.t_radiative = constants.b_wien / (lambda_wien_inner * (
+                1 + (self.v_middle - self.v_inner[0]) / constants.c))
+        else:
+            self.t_radiative = t_radiative
+
+        if dilution_factor is None:
+            self.dilution_factor = 0.5 * (1 - np.sqrt(
+                1 - (self.r_inner[0] ** 2 / self.r_middle ** 2).to(1).value))
```
Comment (ftsamis) [inline tardis/model/base.py]: This is the geometric_w. @wkerzendorf  Do you think we should have it as a property? Would it ever be accessed again except initializing the w?

File: tardis/model/base.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        self.velocity = velocity
+        self.homologous_density = homologous_density
+        self.abundance = abundance
+        self.t_inner = t_inner
+        self.time_explosion = time_explosion
+
+        if t_radiative is None:
+            lambda_wien_inner = constants.b_wien / self.t_inner
+            self.t_radiative = constants.b_wien / (lambda_wien_inner * (
+                1 + (self.v_middle - self.v_inner[0]) / constants.c))
```
Comment (ftsamis) [inline tardis/model/base.py]: This is the initialization of `t_radiative`, taken from the now-legacy model's `_init_t_rad` staticmethod. I think that I should make `self.v_inner[0]` be `self.v_boundary_inner` instead. Is that correct?

File: tardis/model/base.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+from tardis.model import density
+from tardis.util import quantity_linspace, InstanceDescriptorMixin
+
+
+class Radial1DModel(InstanceDescriptorMixin):
+    def __init__(self, velocity, homologous_density, abundance, t_inner,
+                 time_explosion, t_radiative=None, dilution_factor=None,
+                 v_boundary_inner=None, v_boundary_outer=None):
+        self.velocity = velocity
+        self.homologous_density = homologous_density
```
Comment (ftsamis) [inline tardis/model/base.py]: Currently there is no way of accessing the non-adapted to `time_explosion` density array, i.e. `model.density` (which is now the same as `model.homologous_density`) will always return the scaled density. Do we need access to the starting density/density_0/however you call it?

Also, can we just name this density, instead of homologous_density?

File: tardis/model/base.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+            # check for zero volume of designated first cell
+            if np.isclose(self.v_boundary_inner, self.raw_velocity[idx + 1],
+                          atol=1e-8 * u.km / u.s) and (self.v_boundary_inner <=
+                                                           self.raw_velocity[idx + 1]):
+                idx += 1
+            return idx
+
+    @property
+    def v_boundary_outer_index(self):
+        if self.v_boundary_outer >= self.raw_velocity[-1]:
```
Comment (wkerzendorf) [inline tardis/model/base.py]: you are extrapolating here. both boundaries should be within the model and inner < outer. I think we should have a function where we have all these rules in place.

File: tardis/model/density.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+return the density array according to various values of the model, such as the
+velocity, or the time_explosion.
+
+Therefore, changing the velocity array, or the time_explosion in the model
+will automagically change the density array too."""
+
+import numpy as np
+
+
+class Density(object):
```
Comment (wkerzendorf) [inline tardis/model/density.py]: rename to `HomologousDensity`.

File: tardis/model/base.py [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+    def v_boundary_inner(self, value):
+        if value is not None:
+            if value > self.v_boundary_outer:
+                raise ValueError('v_boundary_inner must not be higher than '
+                                 'v_boundary_outer.')
+            if value > self.raw_velocity[-1]:
+                raise ValueError('v_boundary_inner is outside of '
+                                 'the model range.')
+            value = max(value, self.raw_velocity[0])
+        self._v_boundary_inner = value
```
Comment (wkerzendorf) [inline tardis/model/base.py]: `u.Quantity(value, self._v_boundary_inner.unit)`

File: tardis/model/base.py [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+    def __init__(self, density_0, time_0):
+        self.density_0 = density_0
+        self.time_0 = time_0
+
+    def after_time(self, time_explosion):
+        return (self.density_0 * (time_explosion / self.time_0) ** -3).cgs
+
+    @classmethod
+    def from_config(cls, config):
+        d_conf = config.structure.density
```
Comment (wkerzendorf) [inline tardis/model/base.py]:
```python
adjusted_velocity = np.hstack((0, velocity))
v_middle = adjusted_velocity[:-1] * 0.5 + adjusted_velocity[1:] * 0.5
```

Comment (ftsamis) [inline tardis/model/base.py]: Using insert instead of hstack because hstack doesn't preserve the unit.
```
adjusted_velocity = velocity.insert(0, 0)
```

File: tardis/model/base.py (hunk 14 above)
Comment (ftsamis) [inline tardis/model/base.py]: Done.

**Comment by ftsamis** [comment]
Close in favour of #632, so changes are merged with upstream tardis/restructure instead of upstream tardis/master.
